presto-main/src/main/java/com/facebook/presto/execution/buffer/LazyBuffer.java

```diff
@@ -0,0 +1,297 @@
+    {
+        requireNonNull(taskId, "taskId is null");
+        requireNonNull(executor, "executor is null");
+        this.taskInstanceId = requireNonNull(taskInstanceId, "taskInstanceId is null");
+        state = new StateMachine<>(taskId + "-buffer", executor, OPEN, TERMINAL_BUFFER_STATES);
+        this.maxBufferSize = requireNonNull(maxBufferSize, "maxBufferSize is null");
+        checkArgument(maxBufferSize.toBytes() > 0, "maxBufferSize must be at least 1");
+        this.systemMemoryUsageListener = requireNonNull(systemMemoryUsageListener, "systemMemoryUsageListener is null");
+    }
+
+    @Override
+    public void addStateChangeListener(StateChangeListener<BufferState> stateChangeListener)
+    {
+        state.addStateChangeListener(stateChangeListener);
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return 0.0;
+    }
+
+    @Override
+    public SharedBufferInfo getInfo()
+    {
+        OutputBuffer outputBuffer;
+        synchronized (this) {
+            outputBuffer = delegate;
+        }
+
+        if (outputBuffer == null) {
+            //
+            // NOTE: this code must be lock free to not hanging state machine updates
+            //
+            BufferState state = this.state.get();
+
+            return new SharedBufferInfo(
+                    state,
+                    state.canAddBuffers(),
+                    state.canAddPages(),
+                    0,
+                    0,
+                    0,
+                    0,
+                    ImmutableList.of());
+        }
+        return outputBuffer.getInfo();
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
+    {
+        Set<OutputBufferId> abortedBuffers = ImmutableSet.of();
+        List<PendingRead> pendingReads = ImmutableList.of();
+        OutputBuffer outputBuffer;
+        synchronized (this) {
+            if (delegate == null) {
+                delegate = new SharedBuffer(taskInstanceId, state, maxBufferSize, systemMemoryUsageListener);
+
+                // process pending aborts and reads outside of synchronized lock
+                abortedBuffers = ImmutableSet.copyOf(this.abortedBuffers);
+                this.abortedBuffers.clear();
+                pendingReads = ImmutableList.copyOf(this.pendingReads);
+                this.pendingReads.clear();
+            }
+            outputBuffer = delegate;
+        }
+
+        outputBuffer.setOutputBuffers(newOutputBuffers);
+
+        // process pending aborts and reads outside of synchronized lock
+        abortedBuffers.forEach(outputBuffer::abort);
+        for (PendingRead pendingRead : pendingReads) {
+            pendingRead.process(outputBuffer);
+        }
+    }
+
+    @Override
+    public CompletableFuture<BufferResult> get(OutputBufferId bufferId, long startingSequenceId, DataSize maxSize)
```

Comment:
It seems this variable is sometimes called `startingSequencedId`, and sometimes `token`.